Add otel components command

ChrsMark opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

This PR adds the `components` command for `otel` mode as described at https://github.com/elastic/opentelemetry-dev/issues/396.

Implementation is replicated from https://github.com/open-telemetry/opentelemetry-collector/blob/main/otelcol/command_components.go.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

`otel components` command can be used to list the supported components the EDOT includes.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] I have read and understood the [pull request guidelines](https://github.com/elastic/elastic-agent/blob/main/CONTRIBUTING.md#pull-request-guidelines) of this project.
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Elastic Agent.
-->

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

Build Elastic Agent and execute the command:

```console
./build/elastic-agent otel components          
buildinfo:
    command: ./build/elastic-agent
    description: Elastic opentelemetry-collector distribution
    version: 9.0.0
receivers:
    - name: filebeatreceiver
      module: ""
      stability:
        logs: Alpha
        metrics: Undefined
        traces: Undefined
    - name: filelog
      module: ""
      stability:
        logs: Beta
        metrics: Undefined
        traces: Undefined
    - name: hostmetrics
      module: ""
      stability:
        logs: Development
        metrics: Beta
        traces: Undefined
    - name: httpcheck
      module: ""
      stability:
        logs: Undefined
        metrics: Alpha
        traces: Undefined
    - name: jaeger
      module: ""
      stability:
        logs: Undefined
        metrics: Undefined
        traces: Beta
    - name: k8s_cluster
      module: ""
      stability:
        logs: Beta
        metrics: Beta
        traces: Undefined
    - name: k8sobjects
      module: ""
      stability:
        logs: Beta
        metrics: Undefined
        traces: Undefined
    - name: kubeletstats
      module: ""
      stability:
        logs: Undefined
        metrics: Beta
        traces: Undefined
    - name: metricbeatreceiver
      module: ""
      stability:
        logs: Alpha
        metrics: Undefined
        traces: Undefined
    - name: otlp
      module: ""
      stability:
        logs: Beta
        metrics: Stable
        traces: Stable
    - name: prometheus
      module: ""
      stability:
        logs: Undefined
        metrics: Beta
        traces: Undefined
    - name: zipkin
      module: ""
      stability:
        logs: Undefined
        metrics: Undefined
        traces: Beta
processors:
    - name: attributes
      module: ""
      stability:
        logs: Beta
        metrics: Beta
        traces: Beta
    - name: batch
      module: ""
      stability:
        logs: Beta
        metrics: Beta
        traces: Beta
    - name: elasticinframetrics
      module: ""
      stability:
        logs: Undefined
        metrics: Alpha
        traces: Undefined
    - name: elastictrace
      module: ""
      stability:
        logs: Undefined
        metrics: Undefined
        traces: Alpha
    - name: filter
      module: ""
      stability:
        logs: Alpha
        metrics: Alpha
        traces: Alpha
    - name: k8sattributes
      module: ""
      stability:
        logs: Beta
        metrics: Beta
        traces: Beta
    - name: lsminterval
      module: ""
      stability:
        logs: Undefined
        metrics: Development
        traces: Undefined
    - name: memory_limiter
      module: ""
      stability:
        logs: Beta
        metrics: Beta
        traces: Beta
    - name: resource
      module: ""
      stability:
        logs: Beta
        metrics: Beta
        traces: Beta
    - name: resourcedetection
      module: ""
      stability:
        logs: Beta
        metrics: Beta
        traces: Beta
    - name: transform
      module: ""
      stability:
        logs: Alpha
        metrics: Alpha
        traces: Alpha
exporters:
    - name: debug
      module: ""
      stability:
        logs: Development
        metrics: Development
        traces: Development
    - name: elasticsearch
      module: ""
      stability:
        logs: Beta
        metrics: Development
        traces: Beta
    - name: file
      module: ""
      stability:
        logs: Alpha
        metrics: Alpha
        traces: Alpha
    - name: loadbalancing
      module: ""
      stability:
        logs: Beta
        metrics: Development
        traces: Beta
    - name: otlp
      module: ""
      stability:
        logs: Beta
        metrics: Stable
        traces: Stable
    - name: otlphttp
      module: ""
      stability:
        logs: Beta
        metrics: Stable
        traces: Stable
connectors:
    - name: routing
      module: ""
      stability:
        logs-to-logs: Alpha
        logs-to-metrics: Undefined
        logs-to-traces: Undefined
        metrics-to-logs: Undefined
        metrics-to-metrics: Alpha
        metrics-to-traces: Undefined
        traces-to-logs: Undefined
        traces-to-metrics: Undefined
        traces-to-traces: Alpha
    - name: signaltometrics
      module: ""
      stability:
        logs-to-logs: Undefined
        logs-to-metrics: Alpha
        logs-to-traces: Undefined
        metrics-to-logs: Undefined
        metrics-to-metrics: Alpha
        metrics-to-traces: Undefined
        traces-to-logs: Undefined
        traces-to-metrics: Alpha
        traces-to-traces: Undefined
    - name: spanmetrics
      module: ""
      stability:
        logs-to-logs: Undefined
        logs-to-metrics: Undefined
        logs-to-traces: Undefined
        metrics-to-logs: Undefined
        metrics-to-metrics: Undefined
        metrics-to-traces: Undefined
        traces-to-logs: Undefined
        traces-to-metrics: Alpha
        traces-to-traces: Undefined
extensions:
    - name: file_storage
      module: ""
      stability:
        extension: Beta
    - name: health_check
      module: ""
      stability:
        extension: Beta
    - name: memory_limiter
      module: ""
      stability:
        extension: Development
    - name: pprof
      module: ""
      stability:
        extension: Beta
```

## Related issues

- Closes https://github.com/elastic/opentelemetry-dev/issues/396

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes https://github.com/elastic/opentelemetry-dev/issues/396
- Relates #123
- Requires #123
- Superseds #123
-->
-

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

<!-- CI Cheatsheet
Trigger comments:
/test             (Or `buildkite test this|it`) Triggers unit test pipeline
/test extended    (Or `buildkite test extended`) Triggers integration test pipeline

PR labels:
skip-ci           Skips unit and integration tests
skip-it           Skips integration tests
-->

swiatekm reviewed (commented):
Could you add a simple integration test to verify that this actually works?

Also, could we use upstream's implementantion here instead of copying it?

ChrsMark commented:
> Also, could we use upstream's implementantion here instead of copying it?

I didn't find an obvious way to use it directly from `go.opentelemetry.io/collector/otelcol` since it's not exported: https://github.com/open-telemetry/opentelemetry-collector/blob/c14d4f96d69e0053544d5a68863998b2b69a267b/otelcol/command_components.go#L37

I see our `validate` implementation is also copied/re-implemented. Let me know if I miss anything here.

> Could you add a simple integration test to verify that this actually works?

Shall we test against the real EDOT's output? This would mean that every time we include a new component we will need to update that test too.

swiatekm commented:
> > Also, could we use upstream's implementantion here instead of copying it?
> 
> I didn't find an obvious way to use it directly from `go.opentelemetry.io/collector/otelcol` since it's not exported: https://github.com/open-telemetry/opentelemetry-collector/blob/c14d4f96d69e0053544d5a68863998b2b69a267b/otelcol/command_components.go#L37
> 

Allright, let's leave it as is then.

> I see our `validate` implementation is also copied/re-implemented. Let me know if I miss anything here.
> 
> > Could you add a simple integration test to verify that this actually works?
> 
> Shall we test against the real EDOT's output? This would mean that every time we include a new component we will need to update that test too.

I was just thinking about checking if the output looks reasonable, without enumerating the components, but maybe that's a good idea? I'm not sure if we have an integration test that verifies which components are present right now.

elasticmachine commented:
Pinging @elastic/elastic-agent-control-plane (Team:Elastic-Agent-Control-Plane)

ChrsMark commented:
Another thing I see that is missing here is the `module` (see the sample output in the PR description) information. The reason is that we don't populate the modules when we instantiate the collector factories at https://github.com/elastic/elastic-agent/blob/dbfd4471abcd5b9cc8ecf7c6d0be94fd4406c80d/internal/pkg/otel/components.go#L66 like it's done in the [otel builder](https://github.com/open-telemetry/opentelemetry-collector/blob/795f727e05eb29120690d9c5bc0fc5d1c60a9252/cmd/otelcorecol/components.go#L26).

Is this something that we should fix or we can just skip it from the output for now? Any thoughts?

swiatekm commented:
> Another thing I see that is missing here is the `module` (see the sample output in the PR description) information. The reason is that we don't populate the modules when we instantiate the collector factories at
> 
> https://github.com/elastic/elastic-agent/blob/dbfd4471abcd5b9cc8ecf7c6d0be94fd4406c80d/internal/pkg/otel/components.go#L66
> like it's done in the [otel builder](https://github.com/open-telemetry/opentelemetry-collector/blob/795f727e05eb29120690d9c5bc0fc5d1c60a9252/cmd/otelcorecol/components.go#L26).
> 
> Is this something that we should fix or we can just skip it from the output for now? Any thoughts?

I'd skip it for now and open an issue to include it. We should be as close to builder output as possible.

ChrsMark commented:
All right, I have removed the `module` information for now (filed https://github.com/elastic/elastic-agent/issues/6541 for the generic fix) and added a test that checks that all expected components are found. The test can be further enhanced to check components' stability, buildinfo or whatever else we would like to include. For now I started with the bear minimum.

Feel free to take a look when you get the time.

ChrsMark commented:
I wonder if the coverage failure is legit: https://github.com/elastic/elastic-agent/pull/6539#issuecomment-2598613899.

It complains that the new `Components` function is not covered, however it should be since it's called by the added test 🤔 .

jlind23 commented:
@ChrsMark we had similar problem in the past hence the reason why it became a non required check. Feel free to ignore it.

andrzej-stencel commented:
Can we mention the command in the OTel [README](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/otel/README.md)? I think it would help users make sure the command is supported. A single sentence with an example, similar to the one with `validate` command, would be fine, something like:

> Use the `components` command to get the list of components included in the binary:
> 
> ./elastic-agent otel components

ChrsMark commented:
> Can we mention the command in the OTel [README](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/otel/README.md)? I think it would help users make sure the command is supported. A single sentence with an example, similar to the one with `validate` command, would be fine, something like:
> 
> > Use the `components` command to get the list of components included in the binary:
> > ./elastic-agent otel components

Done!

ChrsMark commented:
@andrzej-stencel @michalpristas @swiatekm I have enabled auto-merge but I guess that's not going to work without `SonarQube` test going green, right?
